### PR TITLE
add t5 model

### PR DESCRIPTION
**Pull request by marigoold**
此PR完成了：
复现了和Megatron对齐的T5模型

对齐曲线如下
图片 3
图片 1

### CODE REVIEW

**Comment by marigoold** [comment]
有一个部分需要修改：
`T5ExtendedAttnMask` 是按照libai bert复现的，但是在和Megatron对齐的过程中，发现Megatron只是简单地做了 `.unsqueeze(1)`，和 `T5ExtendedAttnMask` 的实现不同，因此没有用上。所以这部分需要删除还是修改成等价于 `.unsqueeze(1)` 的操作放到模型里面？

File: libai/models/t5_model.py [hunk 1]
```diff
@@ -0,0 +1,338 @@
+import oneflow as flow
```
Comment (L1aoXingyu) [inline libai/models/t5_model.py:1]: 加一下 header

File: libai/models/t5_model.py [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+class T5Embedding(flow.nn.Module):
+    def __init__(
+        self,
+        hidden_size,
+        vocab_size,
+        max_sequence_length,
+        embedding_dropout_prob,
+        init_method=flow.nn.init.xavier_normal_,
+        num_tokentypes=0,
+        fp16=False,
```
Comment (L1aoXingyu) [inline libai/models/t5_model.py:43]: 把这里的 fp16 去掉吧，这里暂时不搞 fp16

File: libai/models/t5_model.py [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+            "attention_probs_dropout_prob": cfg.attention_probs_dropout_prob,
+            "max_position_embeddings": cfg.max_position_embeddings,
+            "num_tokentypes": cfg.num_tokentypes,
+            "initializer_range": cfg.initializer_range,
+            "layernorm_eps": cfg.layernorm_eps,
+            "bias_gelu_fusion": cfg.bias_gelu_fusion,
+            "bias_dropout_fusion": cfg.bias_dropout_fusion,
+            "scale_mask_softmax_fusion": cfg.scale_mask_softmax_fusion,
+            "apply_query_key_layer_scaling": cfg.apply_query_key_layer_scaling,
+            "fp16": cfg.fp16,
```
Comment (L1aoXingyu) [inline libai/models/t5_model.py:233]: 这里也是，去掉 fp16

File: libai/models/t5_model.py [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+            ]
+        )
+
+        decoder_final_layernorm = LayerNorm(
+            (hidden_size, ), eps=layernorm_eps, layer_idx=-1
+        )
+
+        self.decoder = flow.nn.Sequential()
+        self.decoder.add_module('layers', decoder_layers)
+        self.decoder.add_module('final_layernorm', decoder_final_layernorm)
```
Comment (L1aoXingyu) [inline libai/models/t5_model.py:210]: 这个地方这样写有什么好处吗，对比于单独写一个 decoder 和 decoder_final_layernorm

File: libai/models/t5_model.py [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+            encoder_attn_mask,
+            decoder_attn_mask,
+            encoder_decoder_attn_mask,
+            None,
+        )
+
+        lm_loss = self.loss_func(logits, lm_labels, loss_mask)
+        return lm_loss
+
+@GRAPH_REGISTRY.register()
```
Comment (L1aoXingyu) [inline libai/models/t5_model.py:313]: 现在不需要为某个模型构建 graph 了，可以去掉，依赖这个 pr 的合并
https://github.com/Oneflow-Inc/libai/pull/72

**Review by L1aoXingyu — COMMENTED** [review]
可以对接一下 main 分支，看看将模型放在 main 分支下能不能正常跑通

**Comment by dangkai4u** [comment]
> T5ExtendedAttnMask 是按照libai bert复现的，但是在和Megatron对齐的过程中，发现Megatron只是简单地做了 .unsqueeze(1)，和 T5ExtendedAttnMask 的实现不同，因此没有用上。所以这部分需要删除还是修改成等价于 .unsqueeze(1) 的操作放到模型里面？

这个因为megatron构造t5数据时，encoder_padding_mask等是[bsz, seqlen, seqlen]的维度，在bert里是[bsz, seqlen]的维度，因此它的处理和bert不同。我在https://github.com/Oneflow-Inc/libai/blob/add_models/libai/layers/mask_helpers.py 里把所有的attention mask操作都写成了模块，可以用这个替换掉bert和t5中的ExtendedAttnMask

File: libai/models/t5_model.py (hunk 4 above)
Comment (marigoold) [reply libai/models/t5_model.py:210]: 没啥好处，这么写主要是为了和Megatron的参数名称对齐，方便load参数

**Comment by dangkai4u** [comment]
现在写的模型只能训练，推理都没有实现。T5model拆分为T5Embedding、T5Encoder、T5Decoder三个模块。T5Embedding作为公共模块传参到T5Encoder、T5Decoder中，实现参数共享。T5model的forward过程包括两部分，encoder.forward()和decoder.forward()。T5Encoder实现编码器的所有层以及编码器的调用过程，T5Decoder实现解码器的所有层和解码器的调用过程，编码过程没有特殊之处，仿照bert直接传参输出即可，解码过程要考虑增量解码，所以要仿照其他库实现。必须将整个模型拆分开，这样才能在推理阶段，调用model.encoder(xxx)得到编码状态，然后循环调用model.decoder(xxx)逐一解码。

File: libai/models/t5_model.py [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+            # Change position_ids sbp sign: [B, B] -> [S(0), B]
+            position_ids = (
+                self.position_ids[:, :seq_length]
+                .expand_as(input_ids)
+                .to_consistent(sbp=input_ids.sbp)
+            )
+        position_embeddings = self.position_embeddings(position_ids)
+        embeddings = word_embeddings + position_embeddings
+
+        if self.tokentype_embedding is not None:
```
Comment (dangkai4u) [inline libai/models/t5_model.py:102]: T5Embedding里不用加tokentype_embedding

Comment (marigoold) [reply libai/models/t5_model.py:102]: 是的，我看到Megatron里面也是没有的，所以这里应该直接去掉而不是显式把num_tokentypes设置成0吗？

**Comment by dangkai4u** [comment]
大家可以思考一下，现在验证了什么？目前验证了libai的layers使用到模块是正确的，按照目前的方式组装成t5模型和megatron一样，这种组装方式也是正确的。这仅证明训练的正确性，推理的正确性目前无法证明。那么，同样使用这些模块，改变它的组装方式（模型层以及它们的相对顺序不变），是不是也是正确的。以及，在此基础上，加上一些推理相关的，是不是也是正确的。

我不是想说之前我实现的都对，这个工作没有意义。我之前实现的那个pr，最近就发现了不满意的地方，可能也会存在错误，如果延续那个pr，我仍然会改正那些错误。我想表达的是，megatron的实现不友好，所以我们要实现一个更简单易用的库，同时拓展一下oneflow的生态。它的不友好不仅是mpu模块、dataset、trainer，也包括模型，所以我们不能完全对标它的实现。

现在的问题是，megatron实现的模型，将bert、gpt、t5统一为language_model模块，通过add_encoder和add_decoder参数控制。而我们抽象到transformer_layer层，具体的模型由算法工程师开发决定。这里导致参数名称可能不统一，如果想对齐megatron，只能要求模型开发时，也按照它的组装顺序实现。这导致了另一个问题，megatron的推理阶段是通过attention里inference_params这些参数控制的，来减少重复计算。我们之前实现attention时参考huggingface实现，通过past_key_value和use_cache参数控制。如果按照megatron实现，推理无法完成；如果按照huggingface实现，对齐精度时会增加加载模型的工作量。你们可以思考一下之后按照哪种方式实现

File: libai/models/t5_model.py (hunk 6 above)
Comment (dangkai4u) [reply libai/models/t5_model.py:102]: 删掉吧

File: libai/models/t5_model.py [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+                decoder_attn_mask, 
+                encoder_states,
+                encoder_decoder_attn_mask,
+            )
+        decoder_states = self.decoder.final_layernorm(dec_hidden_states)
+        logits = self.lm_head(decoder_states, self.embedding.word_embeddings.weight)
+        return logits
+
+
+class T5Loss(flow.nn.Module):
```
Comment (dangkai4u) [inline libai/models/t5_model.py:268]: 咱们应该重新编写一个ParallelCrossEntropyLoss模块，尽可能和torch的用法一样，例如增加ignore_index参数而非使用loss_mask数组，如果可以加上label smoothing。话说oneflow内置的交叉熵损失，除了sparse_softmax_cross_entropy，其他都不支持分布式计算吗

**Comment by L1aoXingyu** [comment]
我觉得这些工作的目的只是为了有一个标准来衡量一个模型的正确性，不管是用 huggingface 还是 megatron 都是可以的，我们可以选择更简单的那个方式来。

像 NLP pretraining 这种 task 哪怕在小规模数据集上验证也比较麻烦，所以我们找一个标准去对齐 loss，这样发现问题可以及早暴露，而不是一直在写代码，最后发现精度跑不出来，然后无法定位是哪里写错了。

后续 bert 会专门保留一个分支用作 loss 对齐，每次 layers 里面的东西和 bert 模型本身有了优化和修改，都需要在那个分支下做一次 1000 轮训练已验证改动没有精度问题，这个流程已经非常简单且半自动化。

File: libai/models/t5_model.py (hunk 6 above)
Comment (marigoold) [reply libai/models/t5_model.py]: 已修改

File: libai/models/t5_model.py [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+    VocabEmbedding,
+    TransformerLayer,
+    ParallelCrossEntropyLoss,
+)
+from libai.models.build import MODEL_ARCH_REGISTRY, GRAPH_REGISTRY
+from libai.models.utils import init_method_normal, scaled_init_method_normal, GraphBase
+
+from libai.utils import distributed as dist
+
+class T5ExtendedAttnMask(flow.nn.Module):
```
[the remark this answers is not in the document]
Comment (marigoold) [reply libai/models/t5_model.py]: 已修改

File: libai/models/t5_model.py (hunk 3 above)
Comment (marigoold) [reply libai/models/t5_model.py]: 已修改

File: libai/models/t5_model.py (hunk 2 above)
Comment (marigoold) [reply libai/models/t5_model.py]: 已修改

File: libai/models/t5_model.py (hunk 1 above)
Comment (marigoold) [reply libai/models/t5_model.py:16]: 已添加

**Comment by marigoold** [comment]
> > T5ExtendedAttnMask 是按照libai bert复现的，但是在和Megatron对齐的过程中，发现Megatron只是简单地做了 .unsqueeze(1)，和 T5ExtendedAttnMask 的实现不同，因此没有用上。所以这部分需要删除还是修改成等价于 .unsqueeze(1) 的操作放到模型里面？
> 
> 这个因为megatron构造t5数据时，encoder_padding_mask等是[bsz, seqlen, seqlen]的维度，在bert里是[bsz, seqlen]的维度，因此它的处理和bert不同。我在https://github.com/Oneflow-Inc/libai/blob/add_models/libai/layers/mask_helpers.py 里把所有的attention mask操作都写成了模块，可以用这个替换掉bert和t5中的ExtendedAttnMask

这个改完了

File: libai/models/t5_model.py [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+                    scale_mask_softmax_fusion=scale_mask_softmax_fusion,
+                    apply_query_key_layer_scaling=apply_query_key_layer_scaling,
+                    layer_idx=i,
+                )
+                for i in range(hidden_layers)
+            ]
+        )
+
+        encoder_final_layernorm = LayerNorm(
+            (hidden_size, ), eps=layernorm_eps, layer_idx=-1
```
Comment (L1aoXingyu) [inline libai/models/t5_model.py:137]: 这个地方的 layer_idx 在 megatron 里面也是放到最后了吗

**Review by L1aoXingyu — COMMENTED** [review]
我这边大体上没有问题了，在 main 分支上跑通训练，可以正常收敛就可以了

File: libai/models/t5_model.py (hunk 9 above)
Comment (marigoold) [reply libai/models/t5_model.py:137]: https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/model/fused_layer_norm.py#L68-L72
megatron里面没有这个参数，这里我改一下改成encoder所在的层数

Comment (L1aoXingyu) [reply libai/models/t5_model.py:137]: 这个地方的 layer_idx 应该和 encoder_layers 里面的最后一个 layer_idx 一样
